automerge-frontend/src/state_tree/sequence_tree.rs

```diff
@@ -0,0 +1,672 @@
+{
+    pub fn new() -> Self {
+        Self { root_node: None }
+    }
+
+    pub fn len(&self) -> usize {
+        self.root_node.as_ref().map_or(0, |n| n.len())
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn iter(&self) -> Iter<'_, T, B> {
+        Iter {
+            inner: self,
+            index: 0,
+        }
+    }
+
+    pub fn insert(&mut self, mut index: usize, opid: OpId, element: T) {
+        let old_len = self.len();
+        if let Some(root) = self.root_node.as_mut() {
+            #[cfg(debug_assertions)]
+            root.check();
+
+            if root.is_full() {
+                let original_len = root.len();
+                let new_root = SequenceTreeNode::new();
+
+                // move new_root to root position
+                let old_root = mem::replace(root, new_root);
+
+                root.length += old_root.len();
+                root.children.push(old_root);
+                root.split_child(0);
+
+                // after splitting the root has one element and two children, find which child the
+                // index is in
+                let mut i = 0;
+                if root.children[0].len() < index {
+                    i += 1;
+                    index -= root.children[0].len() + 1
+                }
+                assert_eq!(original_len, root.len());
+                root.length += 1;
+                root.children[i].insert_into_non_full_node(index, opid, element)
+            } else {
+                root.insert_into_non_full_node(index, opid, element)
+            }
+        } else {
+            self.root_node = Some(SequenceTreeNode {
+                elements: vec![Box::new((opid, element))],
+                children: Vec::new(),
+                length: 1,
+            })
+        }
+        assert_eq!(self.len(), old_len + 1, "{:#?}", self);
+    }
+
+    pub fn push_back(&mut self, opid: OpId, element: T) {
+        let l = self.len();
+        self.insert(l, opid, element)
+    }
+
+    pub fn get(&self, index: usize) -> Option<(OpId, &T)> {
+        self.root_node.as_ref().and_then(|n| n.get(index))
+    }
+
+    pub fn get_mut(&mut self, index: usize) -> Option<(OpId, &mut T)> {
+        self.root_node.as_mut().and_then(|n| n.get_mut(index))
+    }
+
+    pub fn remove(&mut self, index: usize) -> T {
+        if let Some(root) = self.root_node.as_mut() {
+            #[cfg(debug_assertions)]
+            let len = root.check();
+            let old = root.remove(index);
+
+            if root.elements.is_empty() {
+                if root.is_leaf() {
+                    self.root_node = None;
+                } else {
+                    self.root_node = Some(root.children.remove(0));
+                }
+            }
+
+            #[cfg(debug_assertions)]
+            debug_assert_eq!(len, self.root_node.as_ref().map_or(0, |r| r.check()) + 1);
+            old.1
+        } else {
+            panic!("remove from empty tree")
```

Comment:
Could we return a `Result` of some kind rather than panicking? If that's not possible we should at least document the cases when this can panic I think.

Reply:
I've tried to follow a similar API to `Vec`, which doesn't return an option but panics if you access an element outside of the range. Since this is an internal datastructure (for now) I think it is ok to panic, will document it though.